### PR TITLE
Fix launching Java from the launcher

### PR DESCRIPTION
**Pull request by ImUrX**
- Made Windows not make a cmd appear when locating Java versions
- Fixed current_dir being an empty string which made Windows not like it
- Made JavaVersion a .jar, making it a .class doesnt make things simple
- added some more logging around

### CODE REVIEW

File: gui/src-tauri/Cargo.toml [hunk 1]
```diff
@@ -27,22 +27,7 @@ cfg_aliases = "0.1"
...
-    "window-close",
-    "window-maximize",
-    "window-minimize",
-    "window-set-resizable",
-    "window-set-title",
-    "window-start-dragging",
-    "window-unmaximize",
-    "window-unminimize"
-] }
+tauri = { version = "1.2", features = ["cli", "devtools", "dialog", "fs-all", "path-all", "shell-execute", "window-close", "window-maximize", "window-minimize", "window-set-resizable", "window-set-title", "window-start-dragging", "window-unmaximize", "window-unminimize"] }
```
Comment (ImUrX) [inline gui/src-tauri/Cargo.toml:30]: This is provoked by tauri cli sadly

File: gui/src-tauri/src/main.rs [hunk 2]
```diff
@@ -282,7 +294,19 @@ fn valid_java_paths() -> Vec<(OsString, i32)> {
 				.expect("Failed on executing a Java executable")
 				.code()
 				.map(|code| (p, code))
-				.filter(|(_p, code)| *code >= MINIMUM_JAVA_VERSION)
+				.filter(|(_p, code)| {println!("{}", code); *code >= MINIMUM_JAVA_VERSION})
 		})
 		.collect()
 }
+
+fn executable<W: AsRef<OsStr> + ?Sized>(bin: &W) -> Cow<OsStr> {
```
Comment (TheButlah) [inline gui/src-tauri/src/main.rs:302]: There is no need to 
```
\|/          (__)    
     `\------(oo)
       ||    (__)
       ||w--||     \|/
   \|/
```

We aren't writing performance critical code where a `.clone()` is too expensive. Just return an `OsStr`

**Review by TheButlah — CHANGES_REQUESTED** [review]

File: gui/src-tauri/src/main.rs (hunk 2 above)
Comment (ImUrX) [reply gui/src-tauri/src/main.rs:302]:
```
 __________________ 
< why the hate tho >
 ------------------ 
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

File: gui/src-tauri/src/main.rs [hunk 3]
```diff
@@ -45,7 +51,8 @@ fn is_valid_path(path: &PathBuf) -> bool {
 
 fn get_launch_path(cli: Cli) -> Option<PathBuf> {
 	let mut path = cli.launch_from_path.unwrap_or_default();
-	if is_valid_path(&path) {
+	// The reason from all our problems was path being ""
```
Comment (ButterscotchV) [inline gui/src-tauri/src/main.rs:54]: Could this comment explain what the problem was a little? It's a bit odd mentioning it caused problems but not *what* the problems were.

**Review by ButterscotchV — COMMENTED** [review]
I'll approve under the condition that Butlah approves as well.

File: gui/src-tauri/src/main.rs [hunk 4]
```diff
@@ -282,7 +294,19 @@ fn valid_java_paths() -> Vec<(OsString, i32)> {
 				.expect("Failed on executing a Java executable")
 				.code()
 				.map(|code| (p, code))
-				.filter(|(_p, code)| *code >= MINIMUM_JAVA_VERSION)
+				.filter(|(_p, code)| {println!("{}", code); *code >= MINIMUM_JAVA_VERSION})
```
[the remark this answers is not in the document]
Comment (ImUrX) [reply gui/src-tauri/src/main.rs:297]: thanks for reminding me, i left this on accident

File: gui/src-tauri/src/main.rs (hunk 3 above)
Comment (ImUrX) [reply gui/src-tauri/src/main.rs]: i ended up changing this with an if let

**Comment by ImUrX** [comment]
this is ready for merge

**Review by TheButlah — CHANGES_REQUESTED** [review]
Replacing the slimevr-ui.exe in my installation (everything except gui was installed from the installer), I got:
image

No logs, no popups, no anything. I did enable RUST_LOG=trace

File: gui/src-tauri/src/main.rs [hunk 5]
```diff
@@ -286,3 +299,15 @@ fn valid_java_paths() -> Vec<(OsString, i32)> {
 		})
 		.collect()
 }
+
+fn executable<W: AsRef<OsStr> + ?Sized>(bin: &W) -> OsString {
+	let bin = bin.as_ref();
+	if cfg!(windows) {
+		let mut new = OsString::with_capacity(bin.len() + 4);
```
Comment (TheButlah) [inline gui/src-tauri/src/main.rs:306]: We are working in a codebase that doesn't have rust developers, please focus on keeping the code as simple and readable as possible. just do `OsString::new()` - reserving a particular capacity is not necessary here

**Review by TheButlah — CHANGES_REQUESTED** [review]
I'll give a more thorough review and find the cause of the bug when I get home. I still have only glanced at this.

File: gui/src-tauri/src/main.rs (hunk 5 above)
Comment (ImUrX) [reply gui/src-tauri/src/main.rs:306]: i would do this even in js and it would be less unreadable, ``OsString::with_capacity`` is very verbose tbh

**Review by TheButlah — COMMENTED** [review]
It works! Reviewing code now to see if it can be simplified. Will approve soon otherwise

File: gui/src-tauri/src/main.rs [hunk 6]
```diff
@@ -286,3 +300,15 @@ fn valid_java_paths() -> Vec<(OsString, i32)> {
 		})
 		.collect()
 }
+
+fn executable<W: AsRef<OsStr> + ?Sized>(bin: &W) -> OsString {
```
Comment (TheButlah) [inline gui/src-tauri/src/main.rs:304]: We don't need to be so generic here. Just accept an `&OsStr`. These bounds over complicate things.

Comment (TheButlah) [inline gui/src-tauri/src/main.rs:304]: There is no comment on this function. What does it do? How about naming it `maybe_exe` and say "appends .exe on windows".

Better yet, why is this even a function? Its not like its doing anything particularly reusable. Just have 
```rust
const JAVA_EXE: &OsStr = if cfg!(windows) {
  "java.exe"
} else {
  "java"
};
```

File: gui/src-tauri/src/main.rs [hunk 7]
```diff
@@ -196,6 +206,7 @@ fn main() {
 							.emit_all("server-status", emit_me)
 							.expect("Failed to emit");
 					}
+					println!("Java server receiver died");
```
Comment (TheButlah) [inline gui/src-tauri/src/main.rs:209]:
```suggestion
					log::error!("Java server receiver died");
```

**Review by TheButlah — CHANGES_REQUESTED** [review]

File: gui/src-tauri/src/main.rs [hunk 8]
```diff
@@ -1,11 +1,12 @@
...
-use std::env;
 use std::ffi::{OsStr, OsString};
 use std::io::Write;
+#[cfg(windows)]
+use std::os::windows::process::CommandExt;
 use std::panic;
 use std::path::PathBuf;
 use std::process::{Child, Stdio};
-use std::str::FromStr;
+use std::{env, fs};
```
Comment (TheButlah) [inline gui/src-tauri/src/main.rs:9]:
```suggestion
use std::env;
```

File: gui/src-tauri/src/main.rs (hunk 6 above)
Comment (ImUrX) [reply gui/src-tauri/src/main.rs:304]: didnt know if would be needed for smth else

Comment (TheButlah) [reply gui/src-tauri/src/main.rs:304]: dont add until you know you need it, when the alternative is so much simpler. You can always refactor into a new function

**Review by TheButlah — APPROVED** [review]

**Review by TheButlah — APPROVED** [review]

**Pull request merged**